core/src/main/java/io/questdb/KqueueFileWatcher.java

```diff
@@ -0,0 +1,219 @@
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb;
+
+import io.questdb.cairo.CairoException;
+import io.questdb.std.Files;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Os;
+import io.questdb.std.Unsafe;
+import io.questdb.std.filewatch.FileWatcher;
+import io.questdb.std.str.Path;
+import io.questdb.std.str.Utf8Sequence;
+
+public class KqueueFileWatcher extends FileWatcher {
+    private final int bufferSize;
+    private final int dirFd;
+    private final long eventList;
+    private final long evtDir;
+    private final long evtFile;
+    private final long evtPipe;
+    private final int fileFd;
+    private final int kq;
+    private final int readEndFd;
+    private final int writeEndFd;
+
+    public KqueueFileWatcher(Utf8Sequence filePath, FileEventCallback callback) {
+        super(callback);
+
+        try (Path p = new Path()) {
+
+            p.of(filePath).$();
+            this.fileFd = Files.openRO(p);
+            if (this.fileFd < 0) {
+                throw CairoException.critical(Os.errno()).put("could not open file [path=").put(p).put(']');
+            }
+
+            this.dirFd = Files.openRO(p.parent().$());
+            if (this.dirFd < 0) {
+                int errno = Os.errno();
+                Files.close(this.fileFd);
+                throw CairoException.critical(errno).put("could not open directory [path=").put(p).put(']');
+            }
+        }
+
+        try {
+            this.evtFile = KqueueAccessor.evtAlloc(
+                    this.fileFd,
+                    KqueueAccessor.EVFILT_VNODE,
+                    KqueueAccessor.EV_ADD | KqueueAccessor.EV_CLEAR,
+                    KqueueAccessor.NOTE_DELETE | KqueueAccessor.NOTE_WRITE |
+                            KqueueAccessor.NOTE_ATTRIB | KqueueAccessor.NOTE_EXTEND |
+                            KqueueAccessor.NOTE_LINK | KqueueAccessor.NOTE_RENAME |
+                            KqueueAccessor.NOTE_REVOKE,
+                    0
+            );
+
+            if (this.evtFile == 0) {
+                throw CairoException.critical(Os.errno()).put("could allocate kevent for file");
+            }
+
+            this.evtDir = KqueueAccessor.evtAlloc(
+                    this.dirFd,
+                    KqueueAccessor.EVFILT_VNODE,
+                    KqueueAccessor.EV_ADD | KqueueAccessor.EV_CLEAR,
+                    KqueueAccessor.NOTE_DELETE | KqueueAccessor.NOTE_WRITE |
+                            KqueueAccessor.NOTE_ATTRIB | KqueueAccessor.NOTE_EXTEND |
+                            KqueueAccessor.NOTE_LINK | KqueueAccessor.NOTE_RENAME |
+                            KqueueAccessor.NOTE_REVOKE,
+                    0
+            );
+
+            if (this.evtDir == 0) {
+                throw CairoException.critical(Os.errno()).put("could allocate kevent for directory");
+            }
+
+            long fds = KqueueAccessor.pipe();
```

Comment:
should handle `pipe()` returning a negative value. when either native `pipe()` or `fcntl()` fails.

Reply:
```c
    res = fcntl(fds[0], F_SETFL, O_NONBLOCK);
    if (res < 0) {
        close(fds[0]);
        close(fds[1]);
        return res;
    }
    return (jlong) fds[0] << 32 | (jlong) fds[1];
```

Isn't it technically possible to return a valid negative value for res since the return value is a combination of 2 numbers and we're working with signed numbers?  If so, I can check `readEndFd` and `writeEndFd` separately for negative values after we perfom the bitshift...

Reply:
In theory - it's possible - when `int` has more than 32 bits. 

In practice, `int` has 32 bits on all supported systems. So both `FDs` have the highest bit set to `0`. (because valid FDs are positive). so `(jlong) fds[0] << 32 | (jlong) fds[1];` must have the highest bit set to `0` too -> the returned value cannot be negative.